gdal/frmts/tiledb/tiledbdataset.cpp

```diff
@@ -0,0 +1,724 @@
+/************************************************************************/
+
+class TileDBRasterBand : public GDALPamRasterBand
+{
+    friend class TileDBDataset;
+    protected:
+        TileDBDataset  *poGDS;
+        CPLErr SetBuffer( tiledb::Query& query, GDALDataType eType, void * pImage, int nSize );
+    public:
+        TileDBRasterBand( TileDBDataset *, int );
+        virtual ~TileDBRasterBand();
+        virtual CPLErr IReadBlock( int, int, void * ) override;
+        virtual CPLErr IWriteBlock( int, int, void * ) override;
+        virtual GDALColorInterp GetColorInterpretation() override;
+};
+
+/************************************************************************/
+/*                          TileDBRasterBand()                          */
+/************************************************************************/
+
+TileDBRasterBand::TileDBRasterBand( TileDBDataset *poDSIn, int nBandIn ) :
+    poGDS(poDSIn)
+{
+    poDS = poDSIn;
+    nBand = nBandIn;
+    eDataType = poGDS->eDataType;
+    eAccess = poGDS->eAccess;
+    nRasterXSize = poGDS->nRasterXSize;
+    nRasterYSize = poGDS->nRasterYSize;
+    nBlockXSize = poGDS->nBlockXSize;
+    nBlockYSize = poGDS->nBlockYSize;
+}
+
+/************************************************************************/
+/*                          ~TileDBRasterBand()                         */
+/************************************************************************/
+TileDBRasterBand::~TileDBRasterBand()
+
+{
+    FlushCache();
+}
+
+
+/************************************************************************/
+/*                             SetBuffer()                              */
+/************************************************************************/
+
+CPLErr TileDBRasterBand::SetBuffer( tiledb::Query& query, GDALDataType eType, void * pImage, int nSize )
+{
+ switch (eDataType)
+    {
+        case GDT_Byte:
+            query.set_buffer( TILEDB_VALUES, reinterpret_cast<unsigned char*>( pImage ), nSize );
+            break;
+        case GDT_UInt16:
+            query.set_buffer( TILEDB_VALUES, reinterpret_cast<unsigned short*>( pImage ), nSize );
+            break;
+        case GDT_UInt32:
+            query.set_buffer( TILEDB_VALUES, reinterpret_cast<unsigned int*>( pImage ), nSize );
+            break;
+        case GDT_Int16:
+            query.set_buffer( TILEDB_VALUES, reinterpret_cast<short*>( pImage ), nSize );
+            break;
+        case GDT_Int32:
+            query.set_buffer( TILEDB_VALUES, reinterpret_cast<int*>( pImage ), nSize );
+            break;
+        case GDT_Float32:
+            query.set_buffer( TILEDB_VALUES, reinterpret_cast<float*>( pImage ), nSize );
+            break;
+        case GDT_Float64:
+            query.set_buffer( TILEDB_VALUES, reinterpret_cast<double*>( pImage ), nSize );
+            break;
+        case GDT_CInt16:
+            query.set_buffer( TILEDB_VALUES, reinterpret_cast<short*>( pImage ), nSize );
```

Comment:
For the complex types, I think the buffer size should be `nSize * 2` here right?

Reply:
I am not sure about complex types and TileDB yet and need to add tests. I agree with your comment on `nsize * 2`  but I wonder if the type should be `short**` as it is an array of arrays but that didn't match TileDB types.

Reply:
We assume that the struct type (array of fixed sized arrays) is contiguous in memory for writes, I would expect that to be the same for the PDAL buffer